vocabularies/Analytics.json

```diff
@@ -161,6 +164,102 @@
         "@Validation.ApplicableTerms@Core.Description": "Adding a list of other terms that can be annotated to it.",
         "@Validation.ApplicableTerms": ["Common.Label"]
       }
+    },
+    "AutoExpand": [
+      {
+        "$Kind": "Function",
+        "$EntitySetPath": "InputSet",
+        "$IsBound": true,
+        "@Common.Experimental": true,
+        "@Core.Description": "`$apply` transformation that expands an unnamed leveled hierarchy with custom aggregation of certain properties",
+        "@Core.LongDescription": "Example transformation sequence:\n```\n$apply=filter(Industry eq 'IT')\n/groupby((Country,CountryName,Region,RegionName),\n         filter($these/aggregate(Amount) gt 0 and\n                $these/aggregate(Currency) ne null))\n/concat(\n  Analytics.AutoExpand(\n    Levels=[{\"P\":[\"Country\",\"CountryName\"]},\n            {\"P\":[\"Region\",\"RegionName\"]}],\n    LeafLevel=[\"Segment\",\"Industry\"],\n    Aggregation=[\"Amount\",\"Currency\"],\n    SiblingOrder=[{\"Property\":\"Amount\",\n                   \"Order\":\"desc\"}],\n    ExpandLevels=[{\"Entry\":[\"US\",\"USA\"],\"Levels\":0},\n                  {\"Entry\":[\"DE\",\"Germany\",\"BW\",\"Baden-Württemberg\"],\"Levels\":2}])\n    /concat(aggregate($count) as ResultEntriesCount,\n            skip(20)/top(10)),\n  aggregate(Amount,Currency))\n```\n",
+        "$Parameter": [
+          {
+            "$Name": "InputSet",
+            "$Collection": true,
+            "$Type": "Edm.EntityType",
+            "@Core.Description": "Entity set to be processed"
+          },
+          {
+            "$Name": "Levels",
+            "$Collection": true,
+            "$Type": "Analytics.AutoExpandLevel",
+            "@Core.Description": "Collection of aggregation levels forming the visible part of a leveled hierarchy",
+            "@Core.LongDescription": "Each element in the collection defines the properties that constitute one level.\n            A property must not referenced by more than one level.\n            The first element in the collection defines the property names of the coarsest level,\n            the following elements define the property names of consecutively finer-grained aggregation levels.\n            The function result comprises the leveled hierarchy with these levels in preorder.\n            All referenced properties must be groupable."
+          },
+          {
+            "$Name": "LeafLevel",
+            "$Collection": true,
+            "@Core.Description": "A possibly empty set of property names that constitute, together with the property names from `Levels`, the leaf level of the leveled hierarchy",
```

Comment:
> A possibly empty set of property names

I wonder how that would work. If `LeafLevel` is empty then the last level in `Levels` is not expanded, cannot be expanded and could as well be written into `LeafLevel`?

Reply:
Isn't `LeafLevel` simply the last entry in `Levels`, then?

Reply:
No. There can be several properties used for grouping on the leaf level that are not used for visual grouping / group headers.
The interesting part is that we have three things:
1. properties used for visual grouping
2. properties used additionally for grouping on the leaf level
3. the number of visual groups initially expanded.

I think we had decided somewhen in the past that 3. is not necessary for the AutoExpand function as one can simply as collapsed visual groups to 2. But I wonder whether that decision is still good when introducing ExpandLevels.

Reply:
The leaf level is just the deepest grouping level, therefore we only need the parameter `Levels`, not `LeafLevel`.

In the current definition, nodes not mentioned in `ExpandLevels` are completely expanded. If you want the default to be "expand N levels", we need an additional parameter for that N.

Reply:
Yes, we can do it like that. But the question is how to denote the leaf level. Is it always the last of the Levels? And hence this last level is not and cannot be expanded?

Reply:
Yes.

Reply:
Should it be stated explicitly in the Levels description?

Reply:
For me that is a tautology: The node on the deepest level cannot be expanded.

Reply:
If you look at the FE examples, then on the leaf level there are individual (unaggregated) records. If that is the target state, one could also argue that the last level can still be expanded and below this last level are the unaggregated records.

Reply:
https://github.com/SAP/odata-vocabularies/blob/8c7b80bf767e9b13b706b1aeb24029aae39e9ece/vocabularies/Analytics.xml#L224